src/server.py

```diff
@@ -69,14 +69,80 @@ def response_deconstructor(response):
+
+
+def request_deconstructor(request):
+    """Deconstruct request into basic components.
+    Return method, protocol and validates host header"""
+    request_msg = request
+    urequest = request_msg.decode('utf8')
+    first_pass = urequest.split(CRLF+CRLF, 1)
+    head, body = first_pass
+    head_lines = head.split(CRLF)
+    method, path, protocol = head_lines[0].split()
+    headers = head_lines[1:]
+    headers_split = [header.split(':', 1) for header in headers]
+    headers_dict = {k.lower(): v.strip() for k, v in headers_split}
+    return [method, path, protocol, headers_dict, str(len(first_pass))]
+
+
+class HTTPException(Exception):
+    def __init__(self, code, reason, msg_string):
+        """Initiate an instance of HTTPException
+         with attributes: <code>, <reason>, <msg_string>."""
+        self.code = code
+        self.reason = reason
+        self.msg_string = msg_string
+
+    def response_msg(self):
+        """Generate an error message using <code>,
+         <reason>, <msg_string>. Return a byte-string"""
+        template = u'HTTP/1.1 {} {}\r\n\r\n{}'
+        return template.format(self.code, self.reason, self.msg_string)\
+            .encode('utf-8')
+
 
+def parse_req(request):
+    """Raise an appropriate error for a bad request or
+     return URI for a good request."""
+    request_decon = request_deconstructor(request)
+    if request_decon[0] != u'GET':
+        raise NotImplementedError(u'Method not allowed')
+    elif request_decon[2] != u'HTTP/1.1':
+        raise TypeError(u'Version of HTTP not supported')
+    elif u'host' not in request_decon[3]:
+        raise NameError(u'No <host> in headers')
+    return request_decon[1]
 
-def response_error():
-    """Generate response_error."""
-    response = u'HTTP/1.1 500 Internal_Server_Error\r\n\r\nFailure'
-    return response.encode('utf8')
 
+def response_decision(full_mes):
+    """Based on the raised error, return an appropriate http-exception
+     and return URI = 'None'. Return an ok-response and URI for a
+     good request. Response = byte-string, uri = unicode-string."""
+    uri = u'None'
+    try:
+        parse_req(full_mes)
+    except NotImplementedError:
+        response = HTTPException(u'405', u'Method Not Allowed',
+                                 u'The server supports HTTP/1.1 only.')\
+            .response_msg()
+    except TypeError:
+        response = HTTPException(u'505', u'HTTP Version Not supported',
+                                 u'The server supports HTTP/1.1 only.')\
+            .response_msg()
+    except NameError:
+        response = HTTPException(u'400', u'Bad Request',
+                                 u'No <host> in headers.')\
+            .response_msg()
+    except:
+        response = HTTPException(u'500', u'Internal Server Error',
+                                 u'Something went wrong.')\
+            .response_msg()
+    else:
+        response = response_ok()
+        uri = parse_req(full_mes)
```

Comment:
If you follow the advice above about raising all `HTTPExceptions` with the appropriate messages and codes baked in, this hunk from line 124 - 144 becomes this:
```python
url = "Unknown"
try:
    uri = parse_request(full_mes)
except HTTPException as e:
    response = e.response_msg()
else:
    response = response_ok()
return response, uri
```